sdk/helper/storagepacker/ids2keys.go

```diff
@@ -0,0 +1,106 @@
+package storagepacker
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"github.com/hashicorp/vault/sdk/helper/cryptoutil"
+	"math"
+	"sort"
+	"strings"
+)
+
+type itemRequest struct {
+	// Item ID, provided by client
+	ID string
+
+	// Storage key == hash of ID
+	Key string
+
+	// Stored object, nil if not found
+	Value *Item
+}
+
+func GetItemIDHash(itemID string) string {
+	return hex.EncodeToString(cryptoutil.Blake2b256Hash(itemID))
+}
+
+// Given a list of IDs, calculate their keys generate itemRequests for each.
+func (s *StoragePackerV2) keysForIDs(ids []string) []*itemRequest {
+	requests := make([]*itemRequest, 0, len(ids))
+	for _, id := range ids {
+		requests = append(requests, &itemRequest{
+			ID:    id,
+			Key:   GetItemIDHash(id),
+			Value: nil,
+		})
+	}
+	return requests
+}
+
+// Given a list of Items, calculate their keys generate itemRequests for each.
+func (s *StoragePackerV2) keysForItems(items []*Item) []*itemRequest {
+	requests := make([]*itemRequest, 0, len(items))
+	for _, i := range items {
+		requests = append(requests, &itemRequest{
+			ID:    i.ID,
+			Key:   GetItemIDHash(i.ID),
+			Value: i,
+		})
+	}
+	return requests
+}
+
+// Sort the requests in key order, nondestructively (so we can refer
+// back to the original order.)
+func sortRequests(requests []*itemRequest) []*itemRequest {
+	duplicate := append([]*itemRequest{}, requests...)
+	sort.Slice(duplicate, func(i, j int) bool {
+		return duplicate[i].Key < duplicate[j].Key
+	})
+	return duplicate
+}
+
+func checkForDuplicateIds(ids []string) (bool, string) {
+	idsSeen := make(map[string]bool, len(ids))
+	for _, id := range ids {
+		if _, found := idsSeen[id]; found {
+			return true, id
+		}
+		idsSeen[id] = true
+	}
+	return false, ""
+}
+
+// Return the topmost bucket in the tree for a given key.
+// Used as a defult if the cache is empty or bypassed.
+func (s *StoragePackerV2) firstKey(cacheKey string) (string, error) {
+	rootShardLength := s.BaseBucketBits / 4
+	if len(cacheKey) < rootShardLength {
+		return cacheKey, errors.New("key too short")
+	}
+	return cacheKey[0 : s.BaseBucketBits/4], nil
+}
+
+// Return all topmost buckets in the tree.
+func (s *StoragePackerV2) getAllBaseBucketKeys() []string {
```

Comment:
Would be great to see some test coverage on this. Probably could inject a couple of sane numbers for s.BaseBucketBits and make sure we're getting what we expect out.

Reply:
Well, also I notice that this method is written but never used, so maybe it might make sense to strip it instead? Or perhaps it's here for an intended future use.

Reply:
Oh wait, I see it's in use, just not in an area that's touched by this PR. 
```
$ grep getAllBaseBucketKeys . -R
./sdk/helper/storagepacker/storagepacker_v2.go:	for _, bucketKey := range s.getAllBaseBucketKeys() {
./sdk/helper/storagepacker/ids2keys.go:func (s *StoragePackerV2) getAllBaseBucketKeys() []string {
```

Reply:
storagepacker_v2.go is collapsed (due to size?) in the "files changed" view, but it's part of this PR.

Reply:
:+1: